Framework/MDAlgorithms/src/IntegratePeaksMDHKL.cpp

```diff
@@ -0,0 +1,319 @@
+  auto solidAngleValidator = fluxValidator->clone();
+
+  declareProperty(
+      make_unique<WorkspaceProperty<>>("FluxWorkspace", "", Direction::Input,
+                                       PropertyMode::Optional, fluxValidator),
+      "An optional input workspace containing momentum dependent flux for "
+      "normalization.");
+  declareProperty(make_unique<WorkspaceProperty<>>(
+                      "SolidAngleWorkspace", "", Direction::Input,
+                      PropertyMode::Optional, solidAngleValidator),
+                  "An optional input workspace containing momentum integrated "
+                  "vanadium for normalization "
+                  "(a measure of the solid angle).");
+
+  declareProperty(make_unique<WorkspaceProperty<PeaksWorkspace>>(
+                      "PeaksWorkspace", "", Direction::Input),
+                  "A PeaksWorkspace containing the peaks to integrate.");
+
+  declareProperty(
+      make_unique<WorkspaceProperty<PeaksWorkspace>>("OutputWorkspace", "",
+                                                     Direction::Output),
+      "The output PeaksWorkspace will be a copy of the input PeaksWorkspace "
+      "with the peaks' integrated intensities.");
+}
+
+//----------------------------------------------------------------------------------------------
+/**
+ * Execute the algorithm.
+ */
+void IntegratePeaksMDHKL::exec() {
+  IMDWorkspace_sptr m_inputWS = getProperty("InputWorkspace");
+  Mantid::DataObjects::MDFramesToSpecialCoordinateSystem converter;
+  boost::optional<Mantid::Kernel::SpecialCoordinateSystem> coordinateSystem =
+      converter(m_inputWS.get());
+  if (*coordinateSystem != Mantid::Kernel::SpecialCoordinateSystem::HKL) {
+    std::stringstream errmsg;
+    errmsg << "Input MDWorkspace's coordinate system is not HKL.";
+    throw std::invalid_argument(errmsg.str());
+  }
+
+  /// Peak workspace to integrate
+  PeaksWorkspace_sptr inPeakWS = getProperty("PeaksWorkspace");
+  const double box = getProperty("DeltaHKL");
+  const int gridPts = getProperty("GridPoints");
+  const int neighborPts = getProperty("NeighborPoints");
+  /// Output peaks workspace, create if needed
+  PeaksWorkspace_sptr peakWS = getProperty("OutputWorkspace");
+  if (peakWS != inPeakWS)
+    peakWS = inPeakWS->clone();
+
+  MatrixWorkspace_sptr flux = getProperty("FluxWorkspace");
+  MatrixWorkspace_sptr sa = getProperty("SolidAngleWorkspace");
+
+  IMDEventWorkspace_sptr m_eventWS =
+      boost::dynamic_pointer_cast<IMDEventWorkspace>(m_inputWS);
+  IMDHistoWorkspace_sptr m_histoWS =
+      boost::dynamic_pointer_cast<IMDHistoWorkspace>(m_inputWS);
+  int npeaks = peakWS->getNumberPeaks();
+
+  auto prog = make_unique<Progress>(this, 0.3, 1.0, npeaks);
+  PARALLEL_FOR1(peakWS)
+  for (int i = 0; i < npeaks; i++) {
+    PARALLEL_START_INTERUPT_REGION
+
+    IPeak &p = peakWS->getPeak(i);
+    // round to integer
+    int h = static_cast<int>(std::round(p.getH()));
```

Comment:
👍 for using `std::round`. In this situation though, I think [boost::math::iround](http://www.boost.org/doc/libs/1_61_0/libs/math/doc/html/math_toolkit/rounding/round.html) looks a bit cleaner.